compiler/lib/flow_lexer.ml

```diff
@@ -267,27 +271,71 @@ let decode_identifier =
       trim_start
       (Sedlexing.lexeme_length lexbuf - trim_start - trim_end)
   in
+  let unicode_escape_code lexbuf =
+    let hex = sub_lexeme lexbuf 2 0 in
+    let code = int_of_string ("0x" ^ hex) in
+    code
+  in
+  let codepoint_escape_code lexbuf =
+    let hex = sub_lexeme lexbuf 3 1 in
+    let code = int_of_string ("0x" ^ hex) in
+    code
+  in
+  let is_lead_surrogate c = 0xD800 <= c && c <= 0xDBFF in
+  let is_trail_surrogate c = 0xDC00 <= c && c <= 0xDFFF in
+  let surrogate hi lo = (((hi land 0x3FF) lsl 10) lor (lo land 0x3FF)) + 0x10000 in
+  let trail_surrogate env loc buf lexbuf lead =
```

Comment:
Forgive me I'm not familiar with sedlex, I gather this always fails but I wonder why you still combine the chars and add them to buf if possible.

Reply:
This JS lexer always try to recover. It's the responsibility of the caller to decide what to do in case of errors. This design was inherited from the flow parser that I used as a starting point.
Here, recovering by combining surrogate pair seems to be the most natural behavior.